### PR TITLE
LF: Compile netsed ELets using constant stack

### PR DESCRIPTION
**Pull request by remyhaemmerle-da**
CHANGELOG_BEGIN
CHANGELOG_END

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Typing.scala [hunk 1]
```diff
@@ -763,14 +763,17 @@ private[validation] object Typing {
...
-        introExprVar(vName, typ1).typeOf(body)
-      case Binding(_, _, bound @ _) =>
-        typeOf(body)
-    }
+    private def typeOfLet(bindings: List[Binding], body: Expr): Type =
+      bindings match {
+        case Binding(Some(vName), typ0, expr) :: tail =>
+          checkType(typ0, KStar)
+          val typ1 = checkExpr(expr, typ0)
+          introExprVar(vName, typ1).typeOfLet(tail, body)
```
Comment (cocreature) [inline daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Typing.scala:771]: this doesn’t look stack safe, should we make it stack safe? I don’t expect the list of bindings to grow particularly large but it adds to an already large stack.

File: daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Typing.scala [hunk 2]
```diff
@@ -763,14 +763,17 @@ private[validation] object Typing {
...
+        case Binding(_, _, bound @ _) :: tail =>
```
Comment (cocreature) [inline daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Typing.scala:772]:
```suggestion
        case Binding(None, _, bound @ _) :: tail =>
```

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala [hunk 3]
```diff
@@ -738,10 +738,18 @@ private[lf] final class Compiler(
...
+        case ELet(bindings, body) =>
+          val bounds = bindings.map { case Binding(optBinder, _, bound) =>
+            val se = withOptLabel(optBinder, compile(bound))
+            val p = nextPosition()
+            optBinder.foreach(addExprVar(_, p))
+            se
+          }
+          compile(body) match {
+            case SELet(bounds1, body1) =>
+              SELet(bounds ++ bounds1, body1)
```
Comment (cocreature) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala:750]: do we need to worry about the inner let having references to variables bound in outer let? I forgot if you can have references in multi-lets to previous bindings in daml-lf.

**Review by cocreature — APPROVED** [review]
Nice!

File: daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Typing.scala (hunk 1 above)
Comment (remyhaemmerle-da) [reply daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Typing.scala]: We want, but another time

File: daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Typing.scala (hunk 2 above)
Comment (remyhaemmerle-da) [reply daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Typing.scala]: reverted

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala [hunk 4]
```diff
@@ -734,17 +734,21 @@ private[lf] final class Compiler(
...
+  private[this] def compileELet(
+      eLet0: ELet,
+      bindings: mutable.Builder[SExpr, List[SExpr]] = List.newBuilder,
+  ): SELet = {
+    val binding = eLet0.binding
+    bindings += withOptLabel(binding.binder, compile(binding.bound))
+    val boundPos = nextPosition()
+    binding.binder.foreach(addExprVar(_, boundPos))
+    eLet0.body match {
+      case eLet1: ELet => compileELet(eLet1, bindings)
```
Comment (cocreature) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala:748]: you’re still merging nested lets here which is only valid if we can have references between bindings in a let so my previous comment still holds

Comment (remyhaemmerle-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala:748]: > do we need to worry about the inner let having references to variables bound in outer let? I forgot if you can have references in multi-lets to previous bindings in daml-lf.

We do not have to worry. In Speedy (like in the proto),  the  `Let` construction with multiple binding is a compressed form of nested let, each bound have access to the variable of previous bound. 

In any case you can see the function `let` using in the original implementation was already merging the `SELet`.

Comment (cocreature) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala:748]: great thanks for confirming

**Pull request merged**